File: backend/src/controllers/earnings.ts
import type { Request, Response } from "express";
import { prisma } from "../lib/prisma";

const sumFees = async (doctorId: string, from?: Date) => {
  const where: any = { doctorId, status: "completed" };
  if (from) where.date = { gte: from };
  const result = await prisma.appointment.aggregate({
    where,
    _sum: { fee: true },
    _count: true,
  });
  return { total: result._sum.fee ?? 0, count: result._count };
};

// Doctor's own earnings summary (mobile doctor app Earnings screen)
export const getMyEarnings = async (req: Request, res: Response) => {
  try {
    const doctor = (req as any).user;
    const now = new Date();

    const todayStart = new Date(now);
    todayStart.setHours(0, 0, 0, 0);

    const weekStart = new Date(now);
    weekStart.setDate(weekStart.getDate() - 7);

    const monthStart = new Date(now.getFullYear(), now.getMonth(), 1);
    const yearStart = new Date(now.getFullYear(), 0, 1);

    const [today, week, month, year, allTime, virtual, inPerson, recent] =
      await Promise.all([
        sumFees(doctor.id, todayStart),
        sumFees(doctor.id, weekStart),
        sumFees(doctor.id, monthStart),
        sumFees(doctor.id, yearStart),
        sumFees(doctor.id),
        prisma.appointment.aggregate({
          where: { doctorId: doctor.id, status: "completed", isVirtual: true },
          _sum: { fee: true },
          _count: true,
        }),
        prisma.appointment.aggregate({
          where: {
            doctorId: doctor.id,
            status: "completed",
            isVirtual: false,
          },
          _sum: { fee: true },
          _count: true,
        }),
        prisma.appointment.findMany({
          where: { doctorId: doctor.id, status: "completed" },
          orderBy: { date: "desc" },
          take: 10,
        }),
      ]);

    res.json({
      totalEarnings: allTime.total,
      today: today.total,
      thisWeek: week.total,
      thisMonth: month.total,
      thisYear: year.total,
      availableBalance: allTime.total,
      pendingPayments: 0,
      consultationStats: {
        total: allTime.count,
        virtual: virtual._count,
        inPerson: inPerson._count,
      },
      revenueBreakdown: {
        virtual: virtual._sum.fee ?? 0,
        inPerson: inPerson._sum.fee ?? 0,
      },
      recentTransactions: recent.map((a) => ({
        id: a.id,
        patientName: a.patientName,
        amount: a.fee ?? 0,
        date: a.date,
        isVirtual: a.isVirtual,
      })),
    });
  } catch (error) {
    console.error("Error fetching earnings:", error);
    res.status(500).json({ message: "Server error" });
  }
};

File: mobile/lib/features/doctor/data/earnings.dart
class RecentTransaction {
  RecentTransaction({
    required this.id,
    required this.patientName,
    required this.amount,
    required this.date,
    required this.isVirtual,
  });

  final String id;
  final String patientName;
  final int amount;
  final DateTime date;
  final bool isVirtual;

  factory RecentTransaction.fromJson(Map<String, dynamic> json) {
    return RecentTransaction(
      id: json['id'] as String,
      patientName: json['patientName'] as String? ?? 'Patient',
      amount: json['amount'] as int? ?? 0,
      date: DateTime.parse(json['date'] as String),
      isVirtual: json['isVirtual'] as bool? ?? false,
    );
  }
}

class Earnings {
  Earnings({
    required this.totalEarnings,
    required this.today,
    required this.thisWeek,
    required this.thisMonth,
    required this.thisYear,
    required this.availableBalance,
    required this.pendingPayments,
    required this.consultationTotal,
    required this.consultationVirtual,
    required this.consultationInPerson,
    required this.revenueVirtual,
    required this.revenueInPerson,
    required this.recentTransactions,
  });

  final int totalEarnings;
  final int today;
  final int thisWeek;
  final int thisMonth;
  final int thisYear;
  final int availableBalance;
  final int pendingPayments;
  final int consultationTotal;
  final int consultationVirtual;
  final int consultationInPerson;
  final int revenueVirtual;
  final int revenueInPerson;
  final List<RecentTransaction> recentTransactions;

  factory Earnings.fromJson(Map<String, dynamic> json) {
    final stats = json['consultationStats'] as Map<String, dynamic>? ?? {};
    final breakdown = json['revenueBreakdown'] as Map<String, dynamic>? ?? {};
    final transactions = (json['recentTransactions'] as List? ?? [])
        .map((t) => RecentTransaction.fromJson(t as Map<String, dynamic>))
        .toList();

    return Earnings(
      totalEarnings: json['totalEarnings'] as int? ?? 0,
      today: json['today'] as int? ?? 0,
      thisWeek: json['thisWeek'] as int? ?? 0,
      thisMonth: json['thisMonth'] as int? ?? 0,
      thisYear: json['thisYear'] as int? ?? 0,
      availableBalance: json['availableBalance'] as int? ?? 0,
      pendingPayments: json['pendingPayments'] as int? ?? 0,
      consultationTotal: stats['total'] as int? ?? 0,
      consultationVirtual: stats['virtual'] as int? ?? 0,
      consultationInPerson: stats['inPerson'] as int? ?? 0,
      revenueVirtual: breakdown['virtual'] as int? ?? 0,
      revenueInPerson: breakdown['inPerson'] as int? ?? 0,
      recentTransactions: transactions,
    );
  }
}
